
# [AI 이노베이션 스퀘어] 11기 기본 저녁반
- github: https://github.com/hysKim1/AI_Fundamental
    -  [파이썬 공식문서]https://docs.python.org/ko/3/
    - [Numpy 공식문서]https://numpy.org/devdocs/
    
---
* 본 필기는 23일, 24일 두 수업 내용 합쳐서 정리.

1. [Recommndation](#Recommndation) <br>
      1.1 [추천 시스템](#추천-시스템)<br>
2. [0.README](#0.README)<br>
3. [1.데이터 불러오기 ](데이터-불러오기)      
4. [2.데이터 결합](#3.데이터-결합)<br> 
      3.0 [rename](#rename)<br>
      3.1 [concat](#concat)<br>
      3.2 [append](#append)<br>
      3.3 [merge](#merge)<br>
5. [3.전처리 ](#전처리)            
      5.1  [ drop](#drop)<br>
      5.2  [ indexing](#column-indexing)<br>
      5.3  [ select_dtypes](#select_dtypes)<br>
      5.4  [ to_numeric](#to_numeric)<br>
      5.5  [ max_rows](#max_rows)<br>
      5.6  [ select_dtypes](#select_dtypes)<br>
6. [4.Reshaping ](#6.Reshaping)<br> 
      6.1  [melt](#melt)<br>
      6.2  [stack](#stack)<br>
      6.3  [unstack](#unstack)<br> 
      6.3  [pivot](#pivot)<br>      
7. [결측치](#결측치-값-대체)          
      7.1  [fillna](#fillna )<br>

8. [5.상관분석](#상관분석)          
      8.1  [조건문](#조건문-단축 )<br>
      8.2  [apply](#map,apply,applymap)<br>

## Recommendation
| 추천| 검색 |
|:----|:-----|
|과거 데이터/패턴을 바탕으로 정보를 가져다 줌. |이용자가 원하는 정보를 찾아줌|
 예. Netflix : 과거에 작은 비디오 대여 회사
    - 과거 대여 내역을 바탕으로 이용자에게 신작 추천을 반납기한 쯤에 메시지로 보냄
    - 추천 시스템 도입하여 더많은 사용자에게 세분화된 추천서비스  |

인공지능 기법중 가장 적용하기 쉬움
**현재 갖고 있는 데이터 기반**으로 새로운 데이터 값을 얻어야함


##### 파레토 법칙(Pareto Principle)
- 대부분의 판매는 상위 20%가 전체 매출의 80%를 차지한다(전통적)
    - 대부분의 80%가 경시하는 경향이 있음
- 최근 이것이 깨짐

##### 롱태일의 법칙(Long Tail theory)
- 나머지 80%의 중요성을 강조하며 인터넷과 물류가 발단되면서 시작.
- 예. Amazon - 베스트 셀러 외에 나머지 물건들도 돈이 됨
    - 추천 시스템 도입(대체재,보안재) 
    
##### 문제점: 필터 버블(Filter Bubble)
- 사용자 정보 검색물 바탕으로 특정 정보가 사람들의 Bias를 강화시킴
    - Facebook 실험 
        - 한 그룹은 우울한 피드 노출시켰더니 영향을 미침.
        - 주어진 정보가 사람들의 감정에 영향을 준다는 결과
        - 책추천 :생각 조종자들 

##### 1984(오웰)vs 멋진 신세계 (헉슬리)
 - 오웰 : 정부의 통제
 - 헉슬리 : 통제가 없어 정보의 바다에서 진실이 전달 되지 않음

### 추천 시스템

- Recommender system or recommndation engine

방법

1) 협업 filtering : 비슷한 성향의 사람들(유사도 기반)의 과거의 데이터를 바탕으로 추천

2) 컨텐츠 기반 filtering : 사용자의 데이터를 바탕으로 비슷한 내용을 추천


#### 협업 필터링 
- 비슷한 패턴을 가진 사용자나 항목을 추출하는 기술
    - 통계
       
문제점
1. Cold-start 문제 
    - 기존에 데이터가 없을때 성능이 안좋아지는 문제
2. Long tail문제
    - 대중들이 추천하는것을 더 추천을 할 확률이 높음
        - 예. 대중음악
    - 독특한 취향의 고객 문제 
        - gray sheep : 애매모호한 회색분자
        - black sheep : 특이한경우


#### 콘텐츠 기반 필터링
- 컨텐츠 끼리 유사도를 분석 
- 자체가 어려움
- 콜드 스타트 해결 : 데이터가 없어도 유사도를 기반으로 새로운것 추천 가능
    -  새로운 콘텐츠 추천에 유용
    
####  Hybrid:
- 위의 두개를 결합해서 씀 

알고리즘은 무엇인가
- 2차원의 교차표 어떤 사용자가 어떤 전자기기를 구매하였는지
- 벡터화: 샀는지 여부를 수치로 표현(1,0)하여 각 유저별로 벡터로 표현
    
- collaborative filtering : 자기와 가장 유사한 유저를 찾음
    - 인구 통계학적인 방법 : 성별, 연령, 등 고려
    
### 추천 기법 
  
#### User similarity matrix 
- 내적 이용

1) Cosine similarity : 유사도 판단
   값이 클수록 유사성 높음
   
2) k-최근접 이웃
    - k명의 가장 유사한 사람만 뽑음
      
    - 이미 구매한것 들을 제거한 다른 사용자들이 구매한 다른 것들에 값들에 가중치 부여(그 사람의 유사도를 부여) 
        - 그 중 중복되는 제품들은 가중치가 증가함.
        - 그 중 가장 가중치 큰것을 추천

언제 추천할까?

- 여성 쇼핑몰에는 추천 서비스 많음
- 장바구니
    - 대체재 : 서로 대체 가능
        - 상품 조회 페이지에서는 대체 가능한 다른 기종의 휴대폰 비교하도록 추천
    - 보완재 : 
        - 장바구니에서는 구매가 어느정도 확정되기 떄문에 휴대폰 케이스, 기타 악세사리 추천(보완재)


### 0.README
- 데이터 세트에 대한 설명과 안내
    - 미네소타 대학의 그룹렌즈 리서치 프로젝트에 의해서 모은 데이터
    - 943명의 유저가 1682개의 영화에 대해서 1-5점 사이로 100,000개의 평가를 함.

In [8]:
import pandas as pd

### 1.데이터 불러오기
- `pandas.read_csv('파일위치/파일이름',sep=',')`

- 데이터 info, genre,user, occupation,item 불러오기
    - 각자 구분자, enodcing 확인!
   

   - 데이터 불러올때 해당 파일 위치c:/Users/AI02/Downloads/ml-100k파일에서 u.info파일
   - 첫줄은 컬럼명(header) 아님
   - 구분자 : 띄어쓰기

In [9]:
info = pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.info',header=None, sep=' ')

In [10]:
info #이름지정하지 않아서 기본적으로 숫자값

0        1
0     943    users
1    1682    items
2  100000  ratings

##### 열 이름 지정
- `pandas.read_csv('파일위치/파일이름', names=['열1','열2'...])`



In [11]:
info= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.info',
                  header=None, sep=' ', 
                  names=['number', 'data'])

In [12]:
info

number     data
0     943    users
1    1682    items
2  100000  ratings

In [13]:
genre= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.genre',
                   header=None, sep='|',
                   names=['Genre','Code'])

In [14]:
genre

Genre  Code
0       unknown     0
1        Action     1
2     Adventure     2
3     Animation     3
4    Children's     4
5        Comedy     5
6         Crime     6
7   Documentary     7
8         Drama     8
9       Fantasy     9
10    Film-Noir    10
11       Horror    11
12      Musical    12
13      Mystery    13
14      Romance    14
15       Sci-Fi    15
16     Thriller    16
17          War    17
18      Western    18

user data 에서는 names 를 하게되면 인덱스가 1부터 시작하게 됨
- names, header 가 충동하는 경우가 있음 
- names를 지정하게 되면 user id가 없으면 1부터 시작

In [164]:
user= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.user',
                  header=None, sep='|')
user

0   1  2              3      4
0      1  24  M     technician  85711
1      2  53  F          other  94043
2      3  23  M         writer  32067
3      4  24  M     technician  43537
4      5  33  F          other  15213
..   ...  .. ..            ...    ...
938  939  26  F        student  33319
939  940  32  M  administrator  02215
940  941  20  M        student  97229
941  942  48  F      librarian  78209
942  943  22  M        student  77841

[943 rows x 5 columns]

5개의 열 중에서  4개를 이름 지정하게되면 뒤의 4개만 이름 지정
- 이름 지정 안된 앞줄은 인덱스가 됨
- index_col : 앞의 열을 인덱스로 인식

In [175]:
user= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.user',
                  sep='|',
                  names=['user id','gender','occupation','zip'])

In [176]:
user

user id gender     occupation    zip
1         24      M     technician  85711
2         53      F          other  94043
3         23      M         writer  32067
4         24      M     technician  43537
5         33      F          other  15213
..       ...    ...            ...    ...
939       26      F        student  33319
940       32      M  administrator  02215
941       20      M        student  97229
942       48      F      librarian  78209
943       22      M        student  77841

[943 rows x 4 columns]

In [19]:
occ= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.occupation',
                 header=None, sep='|',
                 names=['occupation'])

In [20]:
occ

occupation
0   administrator
1          artist
2          doctor
3        educator
4        engineer
5   entertainment
6       executive
7      healthcare
8       homemaker
9          lawyer
10      librarian
11      marketing
12           none
13          other
14     programmer
15        retired
16       salesman
17      scientist
18        student
19     technician
20         writer

#### Decode Errror 
구분자가 |, ||가 두개 있음 
UnicodeDecodeError :   encoding   
- `read_csv('파일', **encoding**=  , **names= [   ]** )`
    * readme 참조해서 이름 변경

In [23]:
item= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.item',sep='|',encoding='latin1',
                 names=['movie id' , 'movie title' , 'release date' , 'video release date' ,
             'IMDb URL' ,' unknown' , 'Action' , 'Adventure' , 'Animation' ,
             'Childrens' , 'Comedy' ,'Crime' , 'Documentary' , 'Drama' , 'Fantasy' ,
              'Film-Noir' , 'Horror' , 'Musical' ,' Mystery' ,' Romance' , 'Sci-Fi' ,
             ' Thriller' , 'War' , 'Western' ] )

In [165]:
data= pd.read_csv('C:\\Users\\AI02\\Downloads\\ml-100k\\u.data',sep='\t',
                 names=['user id','item id','rating', 'timestamp'])

In [107]:
data

user id  item id  rating  timestamp
0          196      242       3  881250949
1          186      302       3  891717742
2           22      377       1  878887116
3          244       51       2  880606923
4          166      346       1  886397596
...        ...      ...     ...        ...
99995      880      476       3  880175444
99996      716      204       5  879795543
99997      276     1090       1  874795795
99998       13      225       2  882399156
99999       12      203       3  879959583

[100000 rows x 4 columns]

In [30]:
user

user id gender     occupation    zip
1         24      M     technician  85711
2         53      F          other  94043
3         23      M         writer  32067
4         24      M     technician  43537
5         33      F          other  15213
..       ...    ...            ...    ...
939       26      F        student  33319
940       32      M  administrator  02215
941       20      M        student  97229
942       48      F      librarian  78209
943       22      M        student  77841

[943 rows x 4 columns]

### 2.데이터 결합
데이터들을 합치기 위해서는 **동일한 컬럼명**이어야함
- item 데이터의 movie id를 item id 로 바꾸어 item 데이터의 item id 결합 가능

#### rename
- 변수명 변경
- `rename({ '이전 변수명' : '변경하고 싶은 변수명'}, inplace=False,axis=None)`
    * 변수명에 공백에 있기 때문에 dictionary 방법 이용
    
item, data를 결합하려면 동일한 컬럼명이 필요.
- movie id에서 item id로 변경한것을 그대로 반영

In [40]:
item.rename(columns={'movie id':'item id'})

movie id                                movie title release date  \
0            1                           Toy Story (1995)  01-Jan-1995   
1            2                           GoldenEye (1995)  01-Jan-1995   
2            3                          Four Rooms (1995)  01-Jan-1995   
3            4                          Get Shorty (1995)  01-Jan-1995   
4            5                             Copycat (1995)  01-Jan-1995   
...        ...                                        ...          ...   
1677      1678                          Mat' i syn (1997)  06-Feb-1998   
1678      1679                           B. Monkey (1998)  06-Feb-1998   
1679      1680                       Sliding Doors (1998)  01-Jan-1998   
1680      1681                        You So Crazy (1994)  01-Jan-1994   
1681      1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

      video release date                                           IMDb URL  \
0                    NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...   
1                    NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...   
2                    NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...   
3                    NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...   
4                    NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)   
...                  ...                                                ...   
1677                 NaN  http://us.imdb.com/M/title-exact?Mat%27+i+syn+...   
1678                 NaN  http://us.imdb.com/M/title-exact?B%2E+Monkey+(...   
1679                 NaN      http://us.imdb.com/Title?Sliding+Doors+(1998)   
1680                 NaN  http://us.imdb.com/M/title-exact?You%20So%20Cr...   
1681                 NaN  http://us.imdb.com/M/title-exact?Schrei%20aus%...   

       unknown  Action  Adventure  Animation  Childrens  ...  Fantasy  \
0            0       0          0          1          1  ...        0   
1            0       1          1          0          0  ...        0   
2            0       0          0          0          0  ...        0   
3            0       1          0          0          0  ...        0   
4            0       0          0          0          0  ...        0   
...        ...     ...        ...        ...        ...  ...      ...   
1677         0       0          0          0          0  ...        0   
1678         0       0          0          0          0  ...        0   
1679         0       0          0          0          0  ...        0   
1680         0       0          0          0          0  ...        0   
1681         0       0          0          0          0  ...        0   

      Film-Noir  Horror  Musical   Mystery   Romance  Sci-Fi   Thriller  War  \
0             0       0        0         0         0       0          0    0   
1             0       0        0         0         0       0          1    0   
2             0       0        0         0         0       0          1    0   
3             0       0        0         0         0       0          0    0   
4             0       0        0         0         0       0          1    0   
...         ...     ...      ...       ...       ...     ...        ...  ...   
1677          0       0        0         0         0       0          0    0   
1678          0       0        0         0         1       0          1    0   
1679          0       0        0         0         1       0          0    0   
1680          0       0        0         0         0       0          0    0   
1681          0       0        0         0         0       0          0    0   

      Western  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
1677        0  
1678        0  
1679        0  
1680        0  
1681        0  

[1682 rows x 24 columns]

- movie id에서 item id로 변경한것을 그대로 반영

In [169]:
item.rename({'movie id':'item id'}, axis=1, inplace=True)
item

item id                                movie title release date  \
0           1                           Toy Story (1995)  01-Jan-1995   
1           2                           GoldenEye (1995)  01-Jan-1995   
2           3                          Four Rooms (1995)  01-Jan-1995   
3           4                          Get Shorty (1995)  01-Jan-1995   
4           5                             Copycat (1995)  01-Jan-1995   
...       ...                                        ...          ...   
1677     1678                          Mat' i syn (1997)  06-Feb-1998   
1678     1679                           B. Monkey (1998)  06-Feb-1998   
1679     1680                       Sliding Doors (1998)  01-Jan-1998   
1680     1681                        You So Crazy (1994)  01-Jan-1994   
1681     1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

      video release date                                           IMDb URL  \
0                    NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...   
1                    NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...   
2                    NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...   
3                    NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...   
4                    NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)   
...                  ...                                                ...   
1677                 NaN  http://us.imdb.com/M/title-exact?Mat%27+i+syn+...   
1678                 NaN  http://us.imdb.com/M/title-exact?B%2E+Monkey+(...   
1679                 NaN      http://us.imdb.com/Title?Sliding+Doors+(1998)   
1680                 NaN  http://us.imdb.com/M/title-exact?You%20So%20Cr...   
1681                 NaN  http://us.imdb.com/M/title-exact?Schrei%20aus%...   

       unknown  Action  Adventure  Animation  Childrens  ...  Fantasy  \
0            0       0          0          1          1  ...        0   
1            0       1          1          0          0  ...        0   
2            0       0          0          0          0  ...        0   
3            0       1          0          0          0  ...        0   
4            0       0          0          0          0  ...        0   
...        ...     ...        ...        ...        ...  ...      ...   
1677         0       0          0          0          0  ...        0   
1678         0       0          0          0          0  ...        0   
1679         0       0          0          0          0  ...        0   
1680         0       0          0          0          0  ...        0   
1681         0       0          0          0          0  ...        0   

      Film-Noir  Horror  Musical   Mystery   Romance  Sci-Fi   Thriller  War  \
0             0       0        0         0         0       0          0    0   
1             0       0        0         0         0       0          1    0   
2             0       0        0         0         0       0          1    0   
3             0       0        0         0         0       0          0    0   
4             0       0        0         0         0       0          1    0   
...         ...     ...      ...       ...       ...     ...        ...  ...   
1677          0       0        0         0         0       0          0    0   
1678          0       0        0         0         1       0          1    0   
1679          0       0        0         0         1       0          0    0   
1680          0       0        0         0         0       0          0    0   
1681          0       0        0         0         0       0          0    0   

      Western  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
1677        0  
1678        0  
1679        0  
1680        0  
1681        0  

[1682 rows x 24 columns]

In [170]:
data

user id  item id  rating  timestamp
0          196      242       3  881250949
1          186      302       3  891717742
2           22      377       1  878887116
3          244       51       2  880606923
4          166      346       1  886397596
...        ...      ...     ...        ...
99995      880      476       3  880175444
99996      716      204       5  879795543
99997      276     1090       1  874795795
99998       13      225       2  882399156
99999       12      203       3  879959583

[100000 rows x 4 columns]

####  결합 종류

1. 물리적 결합
    - concat ,append
2.  데이터베이스 스타일 결합
    - join, merge

 - 열 이름에 맞추어 결합

### concat, append
- concat > append
    - concat 이 더 옵션 많음
Pandas 공식 문서 참고 : https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html

### concat
- numpy의 `concatnate( )`은 array 결합에 사용 하기 때문에 사촌지간
- 키 설정 가능 
- axis 설정으로 행,열결합 결정
- `pandas.concat([df1 ,df2 ], axis=0, keys=None) `

* NaN: 자동으로 매칭되는 키 없어 값이 없음.
* 단순 결합 : userid가 두개 존재

In [96]:
pd.concat([user, data],axis=1)  #열결합

user id gender  occupation    zip  user id  item id  rating
0          NaN    NaN         NaN    NaN      196      242       3
1         24.0      M  technician  85711      186      302       3
2         53.0      F       other  94043       22      377       1
3         23.0      M      writer  32067      244       51       2
4         24.0      M  technician  43537      166      346       1
...        ...    ...         ...    ...      ...      ...     ...
99995      NaN    NaN         NaN    NaN      880      476       3
99996      NaN    NaN         NaN    NaN      716      204       5
99997      NaN    NaN         NaN    NaN      276     1090       1
99998      NaN    NaN         NaN    NaN       13      225       2
99999      NaN    NaN         NaN    NaN       12      203       3

[100000 rows x 7 columns]

user id 를 키 설정하니 100,000개 열에서 943개 열로 줄음
- 공통된 값만 갖기 때문

In [97]:
pd.concat([user, data],keys=['user id'])

user id gender     occupation    zip
user id 1         24      M     technician  85711
        2         53      F          other  94043
        3         23      M         writer  32067
        4         24      M     technician  43537
        5         33      F          other  15213
...              ...    ...            ...    ...
        939       26      F        student  33319
        940       32      M  administrator  02215
        941       20      M        student  97229
        942       48      F      librarian  78209
        943       22      M        student  77841

[943 rows x 4 columns]

### append
- `append(df1,df2)`instance method
- concat의 간단한 버전

* merge와 다르게 다른 설정 없이 user id 기반을 결합
* NaN: 자동으로 매칭되는 키 없어 값이 없음.

In [102]:
data.append(user) #위의 merge보다 행 수가 많음.

user id  item id  rating gender     occupation    zip
0        196    242.0     3.0    NaN            NaN    NaN
1        186    302.0     3.0    NaN            NaN    NaN
2         22    377.0     1.0    NaN            NaN    NaN
3        244     51.0     2.0    NaN            NaN    NaN
4        166    346.0     1.0    NaN            NaN    NaN
..       ...      ...     ...    ...            ...    ...
939       26      NaN     NaN      F        student  33319
940       32      NaN     NaN      M  administrator  02215
941       20      NaN     NaN      M        student  97229
942       48      NaN     NaN      F      librarian  78209
943       22      NaN     NaN      M        student  77841

[100943 rows x 6 columns]

## 데이터베이스 스타일 결합

### marge, join
- 관계형 데이터 베이스(RDBMS) 형식
- join > 
- 필요한것을 찾아서 붙여줌
    - How 다양한 join 유형 제공:left outer,right outer, full outer, innr join
        - 중복문제는 
    - 키 :컬럼명이 동일한것을 기준으로 결합
  - `join()`메소드는  인덱스 기준으로 내부에서 merge를 사용하여 결합

### merge
- 데이터 2개를 기준으로 교집합(inner join)인 경우의 값만 반환
- `merge(left, right, how= 'inner', on=None )`

In [90]:
merge=data.merge(user).merge(item)

위의 append,merge와 같은 단순 결합에 비해 적은 77,841행이 있음
        - 위의 데이터와 비교시 NaN이 없음.

In [111]:
data.merge(user)

user id  item id  rating  timestamp gender     occupation    zip
0           22      377       1  878887116      M      executive  66315
1           22      377       1  878887116      F  administrator  60202
2           22      377       1  878887116      M       engineer  02215
3           22      377       1  878887116      M          other  11727
4           22      377       1  878887116      M        student  55109
...        ...      ...     ...        ...    ...            ...    ...
77836       55        7       3  878176047      M          other  45218
77837       55        7       3  878176047      M      scientist  50613
77838       55        7       3  878176047      M      librarian  19102
77839       55        7       3  878176047      M          other  94920
77840       55        7       3  878176047      F          other  80526

[77841 rows x 7 columns]

In [114]:
data.merge(user).info() # int, obj 둘다 있음

<class 'pandas.core.frame.DataFrame'>
Int64Index: 77841 entries, 0 to 77840
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user id     77841 non-null  int64 
 1   item id     77841 non-null  int64 
 2   rating      77841 non-null  int64 
 3   timestamp   77841 non-null  int64 
 4   gender      77841 non-null  object
 5   occupation  77841 non-null  object
 6   zip         77841 non-null  object
dtypes: int64(4), object(3)
memory usage: 4.8+ MB


### 3.전처리
### drop
- 데이터 변경전에 deep copy `copy()` 

-label 은 인덱스 라벨과, 컬럼 명 고려하기 때문에 axis 필수
    - `drop(labels= , axis= ,inplace=False)`
    - `drop(columns=  ,inplace=False)`

In [47]:
data_raw= data.copy()

In [48]:
data.drop(labels='timestamp',axis=1)

user id  item id  rating
0          196      242       3
1          186      302       3
2           22      377       1
3          244       51       2
4          166      346       1
...        ...      ...     ...
99995      880      476       3
99996      716      204       5
99997      276     1090       1
99998       13      225       2
99999       12      203       3

[100000 rows x 3 columns]

In [69]:
data.drop(columns='timestamp',inplace=True) #영구적 변경됨

In [70]:
data

user id  item id  rating
0          196      242       3
1          186      302       3
2           22      377       1
3          244       51       2
4          166      346       1
...        ...      ...     ...
99995      880      476       3
99996      716      204       5
99997      276     1090       1
99998       13      225       2
99999       12      203       3

[100000 rows x 3 columns]

### column indexing
    * 열추출 dictionary, attribute ,fancy indexing
    * 행추출 : loc,iloc, at, iat

In [154]:
data['user id']

0        196
1        186
2         22
3        244
4        166
        ... 
99995    880
99996    716
99997    276
99998     13
99999     12
Name: user id, Length: 100000, dtype: int64

In [155]:
data.user id  # 변수명 선정 잘못됨

SyntaxError: invalid syntax (<ipython-input-155-557056991897>, line 1)

### select_dtypes
- 데이터 타입을 선택
- `select_dtypes(include = )`

In [148]:
data.merge(user).select_dtypes(include='int64') #디폴트

user id  item id  rating  timestamp
0           22      377       1  878887116
1           22      377       1  878887116
2           22      377       1  878887116
3           22      377       1  878887116
4           22      377       1  878887116
...        ...      ...     ...        ...
77836       55        7       3  878176047
77837       55        7       3  878176047
77838       55        7       3  878176047
77839       55        7       3  878176047
77840       55        7       3  878176047

[77841 rows x 4 columns]

zip(우편번호)은 수치가 아닌 object
- series 로 뽑는 순간 영어
- 우편번호는 0부터 시작하므로 숫자로 만들수 없어서 값이 index로 감

In [118]:
data.merge(user).select_dtypes(include='object')

gender     occupation    zip
0          M      executive  66315
1          F  administrator  60202
2          M       engineer  02215
3          M          other  11727
4          M        student  55109
...      ...            ...    ...
77836      M          other  45218
77837      M      scientist  50613
77838      M      librarian  19102
77839      M          other  94920
77840      F          other  80526

[77841 rows x 3 columns]

In [128]:
temp=data.merge(user)['zip'].value_counts().sort_index()
temp  # 0부터 시작하고 알파벳 섞여 있음!

00000    153
01002     34
01040     52
01080     66
01331     57
        ... 
V0R2M     25
V1G4L     25
V3N4P     36
V5A2B     34
Y1A6B     25
Name: zip, Length: 795, dtype: int64

### to_numeric
- 수치형으로 변경

우편번호를 수치형으로 바꾸려 했는데 못바꿈

In [129]:
pd.to_numeric(temp)

00000    153
01002     34
01040     52
01080     66
01331     57
        ... 
V0R2M     25
V1G4L     25
V3N4P     36
V5A2B     34
Y1A6B     25
Name: zip, Length: 795, dtype: int64

In [131]:
pd.options.display.max_rows #bug

60

### max_rows
- n개 까지 보이도록 설정함
- 최신버전의 pandas 이어야만 max_rows 

In [145]:
#!pip install -U pandas --user 

Requirement already up-to-date: pandas in c:\users\ai02\anaconda3\lib\site-packages (1.0.5)


In [146]:
pd.options.display.max_rows=500

In [149]:
print(temp)

00000    153
01002     34
01040     52
01080     66
01331     57
        ... 
V0R2M     25
V1G4L     25
V3N4P     36
V5A2B     34
Y1A6B     25
Name: zip, Length: 795, dtype: int64


### 4.Reshaping
- 각 분석에 맞는 데이터 형태로 변형 필요
    - melt
    - stack
    - unstack
    - pivot


#### long format
- 범용성 좋음( 다양한 곳에서 사용 가능)
    - 행 별 관리 : 추가,삭제 용이
    - melt
    - stack
    
### melt

- `data.melt(id_vars =['index1','index2'])`

In [75]:
data.melt(id_vars =['user id','item id'])

user id  item id variable  value
0          196      242   rating      3
1          186      302   rating      3
2           22      377   rating      1
3          244       51   rating      2
4          166      346   rating      1
...        ...      ...      ...    ...
99995      880      476   rating      3
99996      716      204   rating      5
99997      276     1090   rating      1
99998       13      225   rating      2
99999       12      203   rating      3

[100000 rows x 4 columns]

### stack 
- 100000개의 행, 4개의 열로 변경
- long format !
- SCI : Stack 은 column 을 Index로함.

출처: Pandas 공식홈페이지:https://pandas.pydata.org/pandas-docs/stable/user_guide/reshaping.html

user, item id를 각각 인엑스로 변경하여 stack이용하여 multi-인덱스로 만듦.

In [73]:
data.set_index(['user id','item id']).stack() #

user id  item id        
196      242      rating    3
186      302      rating    3
22       377      rating    1
244      51       rating    2
166      346      rating    1
                           ..
880      476      rating    3
716      204      rating    5
276      1090     rating    1
13       225      rating    2
12       203      rating    3
Length: 100000, dtype: int64

### Wide format
- 추천시스템에서 필요한 데이터 형태(교차표)
    - unstack
    - pivot

### unstack
- wide format 으로  pivot과 비슷한 교차표 형태

- 위의 stack 과 반대로 943 rows × 1682 columns 넓어짐.
    - user, item id를 인덱스로 변경하였으나 unstack으로 최하단 레벨의 item id를 columns으로 변경

In [72]:
data.set_index(['user id','item id']).unstack()

rating                                               ...            \
item id   1    2    3    4    5    6    7    8    9    10    ... 1673 1674   
user id                                                      ...             
1          5.0  3.0  4.0  3.0  3.0  5.0  4.0  1.0  5.0  3.0  ...  NaN  NaN   
2          4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  2.0  ...  NaN  NaN   
3          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
4          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
5          4.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
...        ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
939        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  5.0  NaN  ...  NaN  NaN   
940        NaN  NaN  NaN  2.0  NaN  NaN  4.0  5.0  3.0  NaN  ...  NaN  NaN   
941        5.0  NaN  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN  ...  NaN  NaN   
942        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
943        NaN  5.0  NaN  NaN  NaN  NaN  NaN  NaN  3.0  NaN  ...  NaN  NaN   

                                                 
item id 1675 1676 1677 1678 1679 1680 1681 1682  
user id                                          
1        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  ...  
939      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
940      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
941      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
942      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
943      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[943 rows x 1682 columns]

### pivot 
- index, columns,values 를 하나하나 지정해야함
    - `pivot( index, columns,values )`
- value 값 안보여줌
- Excel의 pivot table 과 동일
- 한 셀에 1개의 값만 할당 가능
    - aggregation function 이 필요한 이유

In [119]:
data.pivot_table(index='user id',columns='item id',values='rating') #unstack과 동일한 표 완성됨

item id  1     2     3     4     5     6     7     8     9     10    ...  \
user id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         4.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   2.0  ...   
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
5         4.0   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
939       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   5.0   NaN  ...   
940       NaN   NaN   NaN   2.0   NaN   NaN   4.0   5.0   3.0   NaN  ...   
941       5.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN  ...   
942       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
943       NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN  ...   

item id  1673  1674  1675  1676  1677  1678  1679  1680  1681  1682  
user id                                                              
1         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
5         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
939       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
940       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
941       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
942       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
943       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[943 rows x 1682 columns]

### 결측치 값 대체
- 이전까지는 dropna, drop 옵션을 이용해서 na 제외시켰는 방법 사용
- 이번에는 결측치를 다른 값으로 대체하는 것
    - 그 값은 데이터 마다 다름

### fillna
- `fillna( 대체하고 싶은 값)`
* NaN 결측치 처리방법 중 하나
    - 모든 결측치를 0으로 대체


In [126]:
basic=basic.fillna(0)
basic

item id  1     2     3     4     5     6     7     8     9     10    ...  \
user id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...   
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
5         4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
939       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0  ...   
940       0.0   0.0   0.0   2.0   0.0   0.0   4.0   5.0   3.0   0.0  ...   
941       5.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0  ...   
942       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
943       0.0   5.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0  ...   

item id  1673  1674  1675  1676  1677  1678  1679  1680  1681  1682  
user id                                                              
1         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
939       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
940       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
941       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
942       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
943       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[943 rows x 1682 columns]

결측치 nan은 값이 nan
- 계산에 영향을 받지 않음

In [194]:
None +1

TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

In [195]:
float('nan')+1

nan

### 5.상관분석
- 표를 만들었으면 가장 유사도 높은 사람들 찾음
- 어느정도 상관이 있는지 분석
  ### corr 
- -1 < x <1사이의 값
   * 0 : 선형적 상관성 없음
   * 1 : 1가까우수록 양의 상관성 강함
   * -1: -1 가까울수록 음의 상관성 강함
- `corr()`
   - **각 열 간의 상관 분석함**
      - `swapaxes,tranpose, T ` 이용하여 열,행 맞바꿈

#### 1> 유저간 상관분석
- 각 유저(user id) 간의 상관분석 해야함

In [134]:
relation = basic.T.corr() #전치한후 상관성 분석
relation

user id       1         2         3         4         5         6         7    \
user id                                                                         
1        1.000000  0.106322 -0.016424  0.021716  0.302592  0.345976  0.320017   
2        0.106322  1.000000  0.082680  0.160323  0.020218  0.197844  0.022886   
3       -0.016424  0.082680  1.000000  0.331674 -0.028322  0.017673 -0.013384   
4        0.021716  0.160323  0.331674  1.000000 -0.002426  0.030206  0.041056   
5        0.302592  0.020218 -0.028322 -0.002426  1.000000  0.153409  0.279677   
...           ...       ...       ...       ...       ...       ...       ...   
939      0.060175  0.204094 -0.001145  0.010890  0.024292  0.059487  0.032591   
940      0.247483  0.190848  0.127539  0.174904  0.182466  0.295949  0.251988   
941      0.115504  0.144177  0.084532  0.140811  0.112571  0.113858  0.008168   
942      0.108984  0.138292  0.102025  0.149575  0.096603  0.266345  0.210390   
943      0.323231  0.054235 -0.023601  0.025367  0.248991  0.194925  0.301997   

user id       8         9         10   ...       934       935       936  \
user id                                ...                                 
1        0.276208  0.038812  0.288491  ...  0.285923  0.068878  0.189272   
2        0.072772  0.143716  0.106861  ...  0.106416  0.288550  0.325233   
3        0.055385  0.043505  0.012080  ... -0.020605  0.018746  0.124174   
4        0.171116  0.089372  0.024018  ...  0.016641  0.019738  0.105135   
5        0.209280  0.025496  0.117166  ...  0.272866  0.039333  0.013999   
...           ...       ...       ...  ...       ...       ...       ...   
939      0.068270  0.021669  0.018157  ...  0.015185  0.416779  0.223173   
940      0.213408  0.096886  0.287583  ...  0.273078  0.074005  0.128224   
941      0.129096  0.132400  0.057182  ...  0.012887  0.189966  0.269370   
942      0.142832  0.070695  0.154295  ...  0.172368  0.043239  0.031500   
943      0.262813  0.044727  0.138188  ...  0.188953  0.176460  0.065226   

user id       937       938       939       940       941       942       943  
user id                                                                        
1        0.146979  0.119444  0.060175  0.247483  0.115504  0.108984  0.323231  
2        0.408466  0.289112  0.204094  0.190848  0.144177  0.138292  0.054235  
3        0.046247  0.088974 -0.001145  0.127539  0.084532  0.102025 -0.023601  
4        0.179598  0.122649  0.010890  0.174904  0.140811  0.149575  0.025367  
5        0.039519  0.085548  0.024292  0.182466  0.112571  0.096603  0.248991  
...           ...       ...       ...       ...       ...       ...       ...  
939      0.207318  0.410470  1.000000  0.049517  0.164739  0.008036  0.100749  
940      0.152222  0.124941  0.049517  1.000000  0.122478  0.221234  0.182613  
941      0.221696  0.297188  0.164739  0.122478  1.000000  0.080130  0.065300  
942      0.101948  0.051432  0.008036  0.221234  0.080130  1.000000  0.127735  
943      0.036712  0.142221  0.100749  0.182613  0.065300  0.127735  1.000000  

[943 rows x 943 columns]

id 123의 상관분석

자기 자신과의 상관 관계는 언제나 1이므로 제외해야함
- 양의 상관성 높은 순으로 정렬

user id 123은 6번,383,10,492,18 인 5 사람과 가장 큰 상관성이 존재

In [138]:
relation[123].sort_values(ascending=False)[1:6]

user id
6      0.414398
383    0.365937
10     0.362127
492    0.361374
18     0.354131
Name: 123, dtype: float64

#### 2> 상관성 높은 유저가 본 영화
유저 123 이랑 상관성 높은 유저 6번의 데이터의 영화 평가 리스트
* 211개의 영화를 평가하였음

In [146]:
data[data['user id']==6]

user id  item id  rating
9            6       86       3
231          6       14       5
612          6       98       5
866          6      463       4
877          6      301       2
...        ...      ...     ...
90756        6      237       2
93521        6      539       2
94111        6      537       4
95462        6      490       5
97068        6      202       3

[211 rows x 3 columns]

유저 6번의 데이터가 본 영화 id 값만 추출

In [144]:
data[data['user id']==6]['item id']

9         86
231       14
612       98
866      463
877      301
        ... 
90756    237
93521    539
94111    537
95462    490
97068    202
Name: item id, Length: 211, dtype: int64

#### values
index 제외하고 값 만 뽑아옴

In [148]:
data[data['user id']==6]['item id'].values

array([ 86,  14,  98, 463, 301, 258,  69, 517,  23, 492, 478, 508, 469,
       357, 143, 275, 294, 221, 532, 211, 180, 477, 297, 500,  95, 475,
       509, 134, 525, 209,  19, 476, 154, 175, 523, 498, 188, 213, 432,
       515, 156, 205, 133,  70, 481, 151, 479, 487, 131, 528,  28, 467,
       136, 117,  32,  15, 135, 286, 513, 125, 484, 458, 199, 183, 499,
       197, 195, 488, 466, 193, 483, 526, 459, 238,  81, 208, 423, 486,
        59, 124,  56, 521, 470, 204, 187, 518, 471, 248, 511, 284, 304,
       189, 506, 269,  87, 318, 276, 534, 524, 100, 480, 536,  21, 200,
       494, 165,  22, 419,  50, 520, 169, 246, 293, 503, 465, 461, 410,
       504, 321, 522,  47, 482, 529, 510, 493, 495, 203, 501, 259, 530,
       485, 137, 367,  64, 174, 505, 507, 516,  13, 310, 186, 512, 191,
       274, 178, 473, 408, 538, 427, 435, 272, 405, 166, 194, 514, 177,
        12, 302, 309,  71, 168, 303, 472, 185,  79, 173, 268, 502, 308,
       153,  89,   7, 192, 527, 519, 491, 317, 489, 257, 127, 42

#### 사용자 함수 정의
위와 같은 단계를 계속 반복해야하므로 재사용을 위해서 function 으로 만듦
- 유저 아이디를 조회하면 해당 유저가 본 영화의 id값만 추출하는 함수 

In [178]:
def user_movies(user_id):
    return data[data['user id']==user_id]['item id'].values

In [150]:
user_movies(383)

array([ 319,  505,  464,  474,  321,  641,  425,  475,  663,  603,   86,
        137,  513,   14,  313,   19,    9,  483,  197,  302,  237,  427,
        639,  135,  272,   81,  268,  504,  357,  100,  654, 1063,  340,
        531,  223,   89,  480,  315,  517,  285,  203,  345,  435,  484,
        514,  134, 1005,  193,  238,  182,  657,  132,  180,  479,  478,
         58,  205,  496,  660,  316,  488,  166,  286,  528,  124,  604,
        736,  188,  185,  213,  200], dtype=int64)

#### 3>  해당 유저가 안본 영화
- 유저 셋이서 지금까지 본 영화 
- 중복값 제거.

유저 383, 3, 10  이 세 사람이서 본 영화(중복 없음)

추천 방식
내가 안본 영화중에서 어떤 영화 추천.
- 높은 유사성 갖은 유저의 높은평점 받은 영화 추천
1. 유저 아이디 입력받으면 그분에게 가장 높은 상관성 갖는사람들 뽑음
2. 뽑고나서 그 k번쨰사람들과의 차집합중에서 평균 평점 높은거 or 곂치는게 많은거



In [176]:
def user_movies(user_id, k):
    return data[data['user id']==user_id]['item id'].values[1:k]

방법 2> 세 사람이서 모두 본 영화

In [184]:
inter=set(user_movies(383))& set(user_movies(6))&set(user_movies(10))  #교집합
inter

{9,
 100,
 124,
 132,
 134,
 135,
 137,
 180,
 182,
 185,
 197,
 200,
 203,
 205,
 223,
 238,
 285,
 286,
 302,
 321,
 340,
 357,
 435,
 474,
 475,
 478,
 479,
 480,
 483,
 484,
 488,
 496,
 504,
 505,
 513,
 531}

### 고급인덱싱 

### isin
데이터프레임에 해당 값이 속해있는지 boolean 값으로 반환 (boolean indexing 방법 이용하는것!)

* 유저 세사람이서 모두 **공통적으로 본 영화(inter) 목록을 본 모든 유저 목록** 추출 

In [187]:
data[data['item id'].isin(inter)]

user id  item id  rating
1          186      302       3
5          298      474       4
50         251      100       4
82         225      480       5
106        250      496       4
...        ...      ...     ...
99943      925      200       2
99949      823      134       5
99957      833      474       5
99962      650      479       5
99999       12      203       3

[7461 rows x 3 columns]

- 그들이 공통적으로 본 영화목록중에서 **유저 6,383,10의 유저 아이디와 일치하는 목록**

In [188]:
data[data['item id'].isin(inter)][data['user id'].isin([6,383,10])]

C:\Users\AI02\anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


user id  item id  rating
621         10      100       5
1218        10      488       5
1318        10      285       5
1598         6      478       4
1967         6      357       4
...        ...      ...     ...
91101      383      124       4
94230       10      484       5
96474      383      185       5
99578       10        9       4
99873      383      200       5

[108 rows x 3 columns]

- 그들이 공통적으로 본 영화목록중에서  유저 6,383,10의 유저 아이디와 일치하는 목록 중에서 **평점 높은 순 정렬**

In [190]:
data[data['item id'].isin(inter)][data['user id'].isin([6,383,10])].sort_values('rating',ascending=False)

C:\Users\AI02\anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


user id  item id  rating
621         10      100       5
43653       10      197       5
35227      383      285       5
33794      383      480       5
33127       10      205       5
...        ...      ...     ...
26144      383      531       3
26605        6      321       3
65286        6      340       2
13225      383      475       2
8397         6      286       2

[108 rows x 3 columns]

- 그들이 공통적으로 본 영화목록중에서  유저 6,383,10의 유저 아이디와 일치하는 목록중에서 **평점 4 이상**

In [202]:
temp=data[data['item id'].isin(inter)][data['user id'].isin([6,383,10])][data.rating > 4]
temp

C:\Users\AI02\anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


user id  item id  rating
621         10      100       5
1218        10      488       5
1318        10      285       5
2110        10      504       5
4174         6      475       5
...        ...      ...     ...
84053      383      286       5
89789       10      479       5
94230       10      484       5
96474      383      185       5
99873      383      200       5

[62 rows x 3 columns]

5> 평점별 가중치 적용
- 유저 123의 상관성 분석에서 자신의 가중치(상관계수)를 넣을 예정

In [205]:
relation.loc[123].sort_values(ascending=False)[1:4]

user id
6      0.414398
383    0.365937
10     0.362127
Name: 123, dtype: float64

* 각 user id에 따라서 각 조선식을 주어 해당 유저인경우 해당 값을 가중치(평점에 곱하여) 할당
    * 유사도 높을수록 가중치가 높아 영향이 더 큼
* 예. temp에 weight이라는 열을 생성해서 유저 6인경우 0.414398,
* 유저 6아니고 유저 383 이면 0.365937, 
* 유저 383 아니라면 나머지(유저 10)는 0.362127
설명은 하단의  [조건문 단축](#조건문-단축), [map,apply,mapapply](#map,apply,mapapply)참고



In [220]:
temp['weight']= temp.apply(
    lambda x:x['rating']*0.414398 if x['user id']==6 else 
   x['rating']* 0.365937 if  x['user id']==383 else
   x['rating']* 0.362127, axis=1)

In [221]:
temp

user id  item id  rating    weight
621         10      100       5  1.810635
1218        10      488       5  1.810635
1318        10      285       5  1.810635
2110        10      504       5  1.810635
4174         6      475       5  2.071990
...        ...      ...     ...       ...
84053      383      286       5  1.829685
89789       10      479       5  1.810635
94230       10      484       5  1.810635
96474      383      185       5  1.829685
99873      383      200       5  1.829685

[62 rows x 4 columns]

내림차순 정렬하니 동점이 많음.

In [224]:
temp.sort_values('weight',ascending=False).iloc[:10]

user id  item id  rating   weight
13592        6      124       5  2.07199
8272         6      135       5  2.07199
43766        6      185       5  2.07199
61566        6      474       5  2.07199
12213        6      238       5  2.07199
11804        6      483       5  2.07199
10342        6      488       5  2.07199
9915         6      197       5  2.07199
19303        6      100       5  2.07199
9486         6      484       5  2.07199

각 영화에 대해서 가중치 가장 높은 5개 

In [232]:
t=temp.groupby('item id').mean().sort_values('weight',ascending=False).iloc[:5]
t

user id  rating    weight
item id                           
475          6.0     5.0  2.071990
474        194.5     5.0  1.950838
238        194.5     5.0  1.950838
137        194.5     5.0  1.950838
100          8.0     5.0  1.941312

해당 영화 475 는 바로 Trainspotting  영화

In [228]:
item[item['item id']==475]

item id           movie title release date  video release date  \
474      475  Trainspotting (1996)  19-Jul-1996                 NaN   

                                          IMDb URL   unknown  Action  \
474  http://us.imdb.com/Title?Trainspotting+(1996)         0       0   

     Adventure  Animation  Childrens  ...  Fantasy  Film-Noir  Horror  \
474          0          0          0  ...        0          0       0   

     Musical   Mystery   Romance  Sci-Fi   Thriller  War  Western  
474        0         0         0       0          0    0        0  

[1 rows x 24 columns]

위의 영화들이 무엇인지 한번에 알아보기 위해서 위의 영화의 index 포함하는 것들만 보여줌

In [233]:
item[item['item id'].isin(t.index)]

item id                                        movie title release date  \
99       100                                       Fargo (1996)  14-Feb-1997   
136      137                                   Big Night (1996)  20-Sep-1996   
237      238                             Raising Arizona (1987)  01-Jan-1987   
473      474  Dr. Strangelove or: How I Learned to Stop Worr...  01-Jan-1963   
474      475                               Trainspotting (1996)  19-Jul-1996   

     video release date                                           IMDb URL  \
99                  NaN    http://us.imdb.com/M/title-exact?Fargo%20(1996)   
136                 NaN  http://us.imdb.com/M/title-exact?Big%20Night%2...   
237                 NaN  http://us.imdb.com/M/title-exact?Raising%20Ari...   
473                 NaN  http://us.imdb.com/M/title-exact?Dr.%20Strange...   
474                 NaN      http://us.imdb.com/Title?Trainspotting+(1996)   

      unknown  Action  Adventure  Animation  Childrens  ...  Fantasy  \
99          0       0          0          0          0  ...        0   
136         0       0          0          0          0  ...        0   
237         0       0          0          0          0  ...        0   
473         0       0          0          0          0  ...        0   
474         0       0          0          0          0  ...        0   

     Film-Noir  Horror  Musical   Mystery   Romance  Sci-Fi   Thriller  War  \
99           0       0        0         0         0       0          1    0   
136          0       0        0         0         0       0          0    0   
237          0       0        0         0         0       0          0    0   
473          0       0        0         0         0       1          0    1   
474          0       0        0         0         0       0          0    0   

     Western  
99         0  
136        0  
237        0  
473        0  
474        0  

[5 rows x 24 columns]

---
위의 흐름 문제때문에 설명을 후반부에 넣었다.


### 조건문 단축
조건절 중첩을 단축

오른쪽에는 식이는데 조건을 여러개 겹침
- 각 조건식을 왼쪽부터 차례대로 조건을 체크해서 조건이 맞는경우에 해당
- 식: 값으로 축약 가능
- 값: 1

In [217]:
a=1
x= '1'if a==1 else 3 if a==2 else 4 if a==3 else '기타 등등'
x

'1'

In [219]:
a=100
x= 1 if a==1 else 3 if a==2 else '3' if a==3 else '기타 등등'
x

'기타 등등'

### map,reduce ,apply
- map
    - function,series, dictionary 모두 지원

-  apply :column-wise/row-wise
     - series, dataframe (df는 axis )
         - series 1차원 , dataframe 2차원
       * series: function
       * dataframe : 막강

- applymap :elementwise
    - function 
    - dataframe 
    
 * series.apply - apply는 함수는 쓸 수 있지만 dictionary는 못 씀
* dataframe.applymap() 가능

### map 
iter 이용하여 series의 값 하나씩 연산하여 그값을 교체

In [200]:
map(lambda x:x+1,[1,2,3,4])   #2,3,4,5가 됨

In [183]:
final.gender.map(lambda x : 1 if x=='M' else 2)

0        1
1        2
2        1
3        1
4        1
        ..
77836    1
77837    1
77838    1
77839    1
77840    2
Name: gender, Length: 77841, dtype: int64

### applymap
- dataframe
- elementwise:  내부의 모든 원소(값)에 적용

In [188]:
final.select_dtypes('int64').applymap(lambda x:x+1)

user id  item id  rating  timestamp
0           23      378       2  878887117
1           23      378       2  878887117
2           23      378       2  878887117
3           23      378       2  878887117
4           23      378       2  878887117
...        ...      ...     ...        ...
77836       56        8       4  878176048
77837       56        8       4  878176048
77838       56        8       4  878176048
77839       56        8       4  878176048
77840       56        8       4  878176048

[77841 rows x 4 columns]

### apply
- fucntion 만 가능
- dictionary 불가
- series,DataFame 둘다 가능
    - DF는 axis 구분 필요( 0: 행, 1: 열)
- `apply( func ,axis=0)`
    - axis=0 각 열별로 계산 적용
    - axis=1 각 행별로 계산 

In [190]:
final=data.merge(user)

In [206]:
final

user id  item id  rating  timestamp gender     occupation    zip
0           22      377       1  878887116      M      executive  66315
1           22      377       1  878887116      F  administrator  60202
2           22      377       1  878887116      M       engineer  02215
3           22      377       1  878887116      M          other  11727
4           22      377       1  878887116      M        student  55109
...        ...      ...     ...        ...    ...            ...    ...
77836       55        7       3  878176047      M          other  45218
77837       55        7       3  878176047      M      scientist  50613
77838       55        7       3  878176047      M      librarian  19102
77839       55        7       3  878176047      M          other  94920
77840       55        7       3  878176047      F          other  80526

[77841 rows x 7 columns]

성별이 남자이면 0 그외(여자이면 ) 1 행별로 적용

In [196]:
final.gender.apply(lambda x: 0 if x=='M' else 1) #function 가능

0        0
1        1
2        0
3        0
4        0
        ..
77836    0
77837    0
77838    0
77839    0
77840    1
Name: gender, Length: 77841, dtype: int64

In [203]:
 temp.apply(lambda x:x['user id']+x['item id'],axis=1)

621      110
1218     498
1318     295
2110     514
4174     481
        ... 
84053    669
89789    489
94230    494
96474    568
99873    583
Length: 62, dtype: int64

In [208]:
final.applyy(lambda x:x['rating']+x['user id'],axis=1)

0        23
1        23
2        23
3        23
4        23
         ..
77836    58
77837    58
77838    58
77839    58
77840    58
Length: 77841, dtype: int64

In [213]:
#전체 데이터 
import numpy as np
final.select_dtypes('int64').apply(np.sum,axis=0) #각 열 의 행의 합

user id             2580777
item id            30515118
rating               273613
timestamp    68677746479191
dtype: int64

In [214]:
final.select_dtypes('int64').apply(np.sum,axis=1)  #각 행의 열의 합.

0        878887516
1        878887516
2        878887516
3        878887516
4        878887516
           ...    
77836    878176112
77837    878176112
77838    878176112
77839    878176112
77840    878176112
Length: 77841, dtype: int64